content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,963 @@
+
+**Note:** Fortran arrays are [one-based](https://fortran-lang.org/en/learn/quickstart/arrays_strings/) by default. So the first element is at index 1. This example reads the previous element as well, so it starts `I` at 2.
+
+`!$OMP PARALLEL DO` is a directive in the form of a Fortran comment (Fortran
+comments start with `!`).`PARALLEL DO` starts a parallel "region" that
+includes the code from `DO` to `ENDDO`.
+
+This tells the compiler that the work in the `DO` loop should be shared amongst
+all the threads available to the program.
+
+Clang has [supported OpenMP](https://blog.llvm.org/2015/05/openmp-support_22.html)
+for many years now. The equivalent C++ code is:
+```C++
+void simple(int n, float *a, float *b)
+{
+    int i;
+
+    #pragma omp parallel for
+    for (i=1; i<n; i++)
+        b[i] = (a[i] + a[i-1]) / 2.0;
+}
+```
+([Compiler Explorer](https://godbolt.org/z/Yh9jb8rKe))
+
+For C++, the directive is in the form of a `#pragma` and attached
+to the `for` loop.
+
+LLVM IR does not know anything about OpenMP specifically, so Clang does all the
+work of converting the intent of the directives into LLVM IR. The output from
+Clang looks like this:
+
+```mlir
+define dso_local void @simple(int, float*, float*)
+  (i32 noundef %n, ptr noundef %a, ptr noundef %b) <...> {
+entry:
+<...>
+  call void (<...>) @__kmpc_fork_call(@simple <...> (.omp_outlined) <...>)
+  ret void
+}
+
+define internal void @simple(int, float*, float*) (.omp_outlined)
+  (ptr <...> %.global_tid.,
+   ptr <...> %.bound_tid.,
+   ptr <...> %n,
+   ptr <...> %b,
+   ptr <...> %a) {
+entry:
+<...>
+  call void @__kmpc_for_static_init_4(<...>)
+<...>
+omp.inner.for.body.i:
+<...>
+omp.loop.exit.i:
+  call void @__kmpc_for_static_fini(<...>)
+<...>
+  ret void
+}
+```
+(output edited for readability)
+
+The body of `simple` no longer does all the work. Instead it uses
+`__kmpc_fork_call` to tell the OpenMP
+[runtime library](https://github.com/llvm/llvm-project/tree/main/openmp)
+to run another function, `simple (.omp_outlined)` to do the work.
+
+This second function is referred to as a "micro task". The runtime library
+splits the work across many instances of the micro task and each time
+the micro task function is called, it gets a different slice of the work.
+
+The number of instances is only known at runtime, and can be controlled with
+settings such as [`OMP_NUM_THREADS`](https://www.openmp.org/spec-html/5.0/openmpse50.html).
+
+The LLVM IR representation of `@simple` includes labels like `omp.loop.exit.i`, but
+these are not specific to OpenMP. They are just normal LLVM IR labels whose name
+includes `omp`.
+
+## Sharing Clang's OpenMP Knowledge
+
+In April 2019 LLVM Flang ("F18" at the time) was approved to join the LLVM
+Project. A month later, it was proposed that LLVM Flang should
+share OpenMP support code with Flang.
```

Comment:
This feels kind of redundant with the previous text in mind. Perhaps something like
```suggestion
Shortly after the time that flang was approved to join the LLVM Project
(in April 2019), it was proposed that LLVM Flang should
share OpenMP support code with Flang.
```

Reply:
I've done a version of this, also I realised it needs to say Flang shares with Clang not Flang with Flang.